atoma-proxy/src/server/mod.rs

```diff
@@ -0,0 +1,220 @@
+                model: model.clone(),
+                result_sender,
+            })
+            .map_err(|err| {
+                error!("Failed to send GetTasksForModel event: {:?}", err);
+                StatusCode::INTERNAL_SERVER_ERROR
+            })?;
+        let tasks = result_receiver
+            .await
+            .map_err(|err| {
+                error!("Failed to receive GetTasksForModel result: {:?}", err);
+                StatusCode::INTERNAL_SERVER_ERROR
+            })?
+            .map_err(|err| {
+                error!("Failed to get GetTasksForModel result: {:?}", err);
+                StatusCode::INTERNAL_SERVER_ERROR
+            })?;
+        if tasks.is_empty() {
+            error!("No tasks found for model {}", model);
+            return Err(StatusCode::NOT_FOUND);
+        }
+        state
+            .sui
+            .write()
+            .await
+            .acquire_new_stack_entry(tasks[0].task_small_id as u64, 1000, 100)
+            .await
+            .map_err(|err| {
+                error!("Failed to acquire new stack entry: {:?}", err);
+                StatusCode::INTERNAL_SERVER_ERROR
+            })?
+    } else {
+        stacks[0].selected_node_id
+    };
+
+    let (result_sender, result_receiver) = oneshot::channel();
+    state
+        .state_manager_sender
+        .send(AtomaAtomaStateManagerEvent::GetNodePublicAddress {
+            node_small_id: selected_node_id,
+            result_sender,
+        })
+        .map_err(|err| {
+            error!("Failed to send GetNodePublicAddress event: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?;
+
+    let node_address = result_receiver
+        .await
+        .map_err(|err| {
+            error!("Failed to receive GetNodePublicAddress result: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?
+        .map_err(|err| {
+            error!("Failed to get GetNodePublicAddress result: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?
+        .ok_or_else(|| {
+            error!("No node address found for node {}", selected_node_id);
+            StatusCode::NOT_FOUND
+        })?;
+
+    state
+        .sui
+        .write()
+        .await
+        .send_openai_api_request(payload, node_address)
+        .await
+        .map_err(|err| {
+            error!("Failed to send OpenAI API request: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })
+        .map(|response| Json(response))
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct NodePublicIpAssignment {
+    node_small_id: u64,
+    public_address: String,
+}
+
+#[axum::debug_handler]
+pub async fn node_public_address_registration(
+    State(state): State<ProxyState>,
+    Json(payload): Json<NodePublicIpAssignment>,
+) -> Result<Json<Value>, StatusCode> {
+    state
+        .state_manager_sender
+        .send(AtomaAtomaStateManagerEvent::UpdateNodePublicAddress {
```

Comment:
This is called the first time a node registers, right ? Why then have a `UpdateNodePublicAddress` handler ?

Reply:
because the address can change, and then the node have to update it.
Renamed to `upsert` that's more appropriate naming